From 7f69b99091de07ab798fd7846e14fb54772e881c Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Tue, 31 Mar 2015 22:53:20 -0700
Subject: [PATCH] power: dc_xpwr_battery: retry reading battery temperature

For values that span two registers (in this case, battery
temperature), sometimes the PMIC would incorrectly return
zero for value in second register. So if the raw battery
temperature ADC value is zero, try again to get the correct
value. Since NTC thermistor is used, value being actual zero
would mean that the battery probably should have already
exploded.

Change-Id: I0254a18e45390d65fb5b196de6b69002b8f5ffc4
Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 drivers/power/dc_xpwr_battery.c | 27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/drivers/power/dc_xpwr_battery.c b/drivers/power/dc_xpwr_battery.c
index e49c6dc..74d2c9b 100644
--- a/drivers/power/dc_xpwr_battery.c
+++ b/drivers/power/dc_xpwr_battery.c
@@ -570,11 +570,30 @@ cur_read_fail:
 
 static int pmic_fg_get_btemp(struct pmic_fg_info *info, int *btemp)
 {
-	int ret, raw_val;
+	int ret, raw_val, cnt = 5;
 
-	ret = pmic_read_adc_val("THERMAL", "BATTEMP", &raw_val, info);
-	if (ret < 0)
-		goto btemp_read_fail;
+	/*
+	 * QUIRK: for values that span two registers, sometimes
+	 * the PMIC would incorrectly return zero for value in
+	 * second register. So if the raw battery temperature
+	 * ADC is zero, try again to get the correct value.
+	 *
+	 * Since NTC thermistor is used, value being actual zero
+	 * would mean that the battery probably should have
+	 * already exploded. ^o^
+	 * The over-temp IRQ would have already fired so
+	 * no need to worry about zero value.
+	 */
+	while (cnt--) {
+		ret = pmic_read_adc_val("THERMAL", "BATTEMP", &raw_val, info);
+		if (ret < 0)
+			goto btemp_read_fail;
+
+		if (raw_val)
+			break;
+		else
+			dev_warn(&info->pdev->dev, "bogus batt temp raw ADC value zero\n");
+	}
 
 	/*
 	 * Convert the TS pin ADC codes in to 10's of Kohms
-- 
1.9.1

